File: slave-node/android/app/src/main/java/com/example/socks5proxy/Socks5Server.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.example.socks5proxy;

import io.netty.bootstrap.Bootstrap;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.*;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
import io.netty.handler.codec.LengthFieldPrepender;
import io.netty.handler.logging.LogLevel;
import io.netty.handler.logging.LoggingHandler;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.util.CharsetUtil;

import java.nio.charset.StandardCharsets;
import java.util.logging.Logger;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Inet6Address;

import java.net.SocketAddress;
import java.net.UnknownHostException;
import java.util.logging.Level;

/**
 * @author KDark
 */
enum Addr {
    V4,
    V6,
    DOMAIN
}

public class Socks5Server {

    private static final Logger log = Logger.getLogger(Socks5Server.class.getName());
//    private static final byte[] MAGIC_FLAG = {0x37, 0x37};

    public static void main(String[] args) throws InterruptedException {
        // Set up the server
        EventLoopGroup bossGroup = new NioEventLoopGroup();
        EventLoopGroup workerGroup = new NioEventLoopGroup();
        try {
            ServerBootstrap bootstrap = new ServerBootstrap();
            bootstrap.group(bossGroup, workerGroup)
                    .channel(NioServerSocketChannel.class)
                    .childHandler(new Socks5ServerInitializer());

            ChannelFuture future = bootstrap.bind(1080).sync();
            future.channel().closeFuture().sync();
        } finally {
            bossGroup.shutdownGracefully();
            workerGroup.shutdownGracefully();
        }
    }

    static class Socks5ServerInitializer extends ChannelInitializer<SocketChannel> {

        @Override
        protected void initChannel(SocketChannel ch) {
            ChannelPipeline pipeline = ch.pipeline();
            pipeline.addLast(new LoggingHandler(LogLevel.INFO));
            pipeline.addLast(new LengthFieldBasedFrameDecoder(1024, 0, 2, 0, 2));
            pipeline.addLast(new LengthFieldPrepender(2));
            pipeline.addLast(new Socks5ServerHandler());
        }
    }

    static class Socks5ServerHandler extends SimpleChannelInboundHandler<ByteBuf> {

        private enum State {
            HANDSHAKE,
            HANDLING,
            RECEIVE_4_BYTES
        }

        private static final int SOCKS_VERSION = 5;
        private State currentState = State.HANDSHAKE;

        @Override
        public void channelReadComplete(ChannelHandlerContext ctx) {
            ctx.flush();
        }

        @Override
        protected void channelRead0(ChannelHandlerContext ctx, ByteBuf msg) throws Exception {
            switch (currentState) {
                case HANDSHAKE: {
//                    int length = msg.readableBytes();
                    // Read the initial SOCKS5 handshake
                    if (msg.readableBytes() < 2) {
                        return;
                    }

                    byte version = msg.readByte();
//                    byte nmethods = msg.readByte();

                    if (version != SOCKS_VERSION) {
                        return;
                    }
                    // Respond with method selection
                    ctx.writeAndFlush(Unpooled.copiedBuffer(new byte[]{SOCKS_VERSION, 0}));
                    currentState = State.HANDLING;
                }
                case HANDLING: {
                    if (msg.readableBytes() < 4) {
                        return;
                    }
                    byte version = msg.readByte();
                    if (version != SOCKS_VERSION) {
                        log.warning("Unsupported command: " + version);
                        ctx.close();
                        return;
                    }

                    byte cmd = msg.readByte();
                    if (cmd != 1) {
                        log.warning("Unsupported command: " + cmd);
                        ctx.close();
                        return;
                    }

                    byte reserved = msg.readByte();
                    byte addrType = msg.readByte();

                    InetSocketAddress remoteAddress;
                    int port = 0;
                    switch (addrType) {
                        case 0x01: // IPv4
                            byte[] ipv4 = new byte[4];
                            msg.readBytes(ipv4);
                            port = msg.readUnsignedShort(); // Correctly read port as unsigned short
                            remoteAddress = new InetSocketAddress(InetAddress.getByAddress(ipv4), port);
                            break;
                        case 0x04: // IPv6
                            byte[] ipv6 = new byte[16];
                            msg.readBytes(ipv6);
                            port = msg.readUnsignedShort(); // Correctly read port as unsigned short
                            remoteAddress = new InetSocketAddress(Inet6Address.getByAddress(ipv6), port);
                            break;
                        case 0x03: // Domain name
                            byte domainLength = msg.readByte();
                            byte[] domain = new byte[domainLength];
                            msg.readBytes(domain);
                            port = msg.readUnsignedShort(); // Correctly read port as unsigned short
                            remoteAddress = new InetSocketAddress(new String(domain, CharsetUtil.UTF_8), port);
                            break;
                        default:
                            log.warning("Unknown address type: " + addrType);
                            ctx.close();
                            return;
                    }
                    ChannelPipeline pipeline = ctx.pipeline();
                    TcpTransferHandler transferHandler = new TcpTransferHandler(remoteAddress.getHostString(), Addr.V4, port);
                    pipeline.addLast(transferHandler);

                    pipeline.remove(this);
                    transferHandler.channelActive(ctx);

                    currentState = State.HANDLING;
                }
            }
        }
    }
}


class TcpTransferHandler extends ChannelInboundHandlerAdapter {

    private final String address;
    private final Addr addr;
    private final int port;

    public TcpTransferHandler(String address, Addr addr, int port) {
        this.address = address;
        this.addr = addr;
        this.port = port;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) {
        ctx.channel().eventLoop().execute(() -> {
            try {
                connectToRemote(ctx);
            } catch (UnknownHostException ex) {
                Logger.getLogger(TcpTransferHandler.class.getName()).log(Level.SEVERE, null, ex);
            }
        });
    }

    private void connectToRemote(ChannelHandlerContext ctx) throws UnknownHostException, UnknownHostException {
        System.out.println("Proxy connect to " + address);

        Bootstrap bootstrap = new Bootstrap();
        bootstrap.group(new NioEventLoopGroup())
                .channel(NioSocketChannel.class)
                .handler(new LoggingHandler(LogLevel.INFO))
                .handler(new ChannelInitializer<SocketChannel>() {
                    @Override
                    protected void initChannel(SocketChannel ch) {
                        // Optionally add more handlers
                    }
                });

        ChannelFuture connectFuture;
        switch (addr) {
            case V4:
                connectFuture = bootstrap.connect(new InetSocketAddress(address, port));
                break;
            case V6: {
                Inet6Address ipv6Address = null;
                try {
                    ipv6Address = (Inet6Address) Inet6Address.getByName(address);
                } catch (UnknownHostException ex) {
                    Logger.getLogger(TcpTransferHandler.class.getName()).log(Level.SEVERE, null, ex);
                }
                connectFuture = bootstrap.connect(new InetSocketAddress(ipv6Address, port));
                break;
            }

            case DOMAIN:
                connectFuture = bootstrap.connect(new InetSocketAddress(address, port));
                break;
            default: {
                System.out.println("Unsupported address type");
                return;
            }
        }

        connectFuture.addListener((ChannelFutureListener) future -> {
            if (!future.isSuccess()) {
                System.out.println("Connection to " + address + " failed");
                ctx.close();
                return;
            }

            System.out.println("Connected to " + address);
            SocketChannel remoteChannel = (SocketChannel) future.channel();
            remoteChannel.config().setKeepAlive(true);
//            remoteChannel.config().setOption(ChannelOption.SO_KEEPALIVE, true);

            sendSocks5Reply(ctx, remoteChannel.localAddress());

            // Start the data relay between client and remote server
            relayTraffic(ctx, remoteChannel);
        });
    }

    private void sendSocks5Reply(ChannelHandlerContext ctx, SocketAddress localAddress) {
        ByteBuf reply = Unpooled.buffer();
        reply.writeByte(5);  // SOCKS version
        reply.writeByte(0);  // Connection succeeded
        reply.writeByte(0);  // Reserved

        if (localAddress instanceof InetSocketAddress) {
            InetSocketAddress inetAddress = (InetSocketAddress) localAddress;
            if (inetAddress.getAddress() instanceof java.net.Inet4Address) {
                reply.writeByte(1);  // IPv4
                reply.writeBytes(inetAddress.getAddress().getAddress());
            } else if (inetAddress.getAddress() instanceof Inet6Address) {
                reply.writeByte(4);  // IPv6
                reply.writeBytes(inetAddress.getAddress().getAddress());
            } else {
                String host = inetAddress.getHostName();
                reply.writeByte(3);  // Domain name
                reply.writeByte(host.length());
                reply.writeBytes(host.getBytes(StandardCharsets.UTF_8));
            }
            reply.writeShort(inetAddress.getPort());
        }

        ctx.writeAndFlush(reply);
    }

    private void relayTraffic(ChannelHandlerContext ctx, SocketChannel remoteChannel) {
        ctx.pipeline().addLast(new RelayHandler(remoteChannel));
        remoteChannel.pipeline().addLast(new RelayHandler(ctx.channel()));
    }

    private static class RelayHandler extends ChannelInboundHandlerAdapter {
        private final Channel relayChannel;

        RelayHandler(Channel relayChannel) {
            this.relayChannel = relayChannel;
        }

        @Override
        public void channelRead(ChannelHandlerContext ctx, Object msg) {
            relayChannel.writeAndFlush(msg);
        }

        @Override
        public void channelInactive(ChannelHandlerContext ctx) {
            if (relayChannel.isActive()) {
                relayChannel.close();
            }
        }

        @Override
        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
            cause.printStackTrace();
            ctx.close();
        }
    }
}

File: slave-node/socks_five_test/lib/socks5_server.dart
import 'dart:io';
import 'package:logging/logging.dart';
import 'dart:typed_data';
import 'dart:convert';
import 'dart:async';

final Logger log = Logger('Socks5Proxy');

// Enum for address types
enum Addr { v4, v6, domain }

// Handler for SOCKS5 server connections
class Socks5ServerHandler {
  final Socket client;
  Socket? remoteSocket;

  States currentState = States.handshake;

  Socks5ServerHandler(this.client);

  void start() {
    client.listen((data) {
      processData(data);
    }, onDone: () {
      log.info('Client disconnected');
      client.close();
      remoteSocket?.destroy();
    }, onError: (error) {
      log.severe('Client error: $error');
      client.close();
      remoteSocket?.destroy();
    });
  }

  void startProxy() {
    //TODO: when remoteSocket is null it wont listen here. that might be not so good. please double check
    remoteSocket?.listen((data) {
      client.add(data);
    }, onDone: () {
      client.destroy();
    });
  }

  void processData(List<int> data) async {
    switch (currentState) {
      case States.handshake:
        print(data.length);
        if (data.length < 2) return;
        final version = data[0];
        if (version != 5) return;
        // Respond with method selection
        Uint8List temp = Uint8List.fromList([5, 0]);
        client.add(temp); // SOCKS5 version and method (no authentication)
        currentState = States.handling;
        break;

      case States.handling:
        print(data.length);
        if (data.length < 4) return;
        final version = data[0];
        if (version != 5) {
          client.close();
          return;
        }

        final cmd = data[1];
        if (cmd != 1) {
          log.warning('Unsupported command: $cmd');
          client.close();
          return;
        }
        final reserved = data[2]; // Reserved byte
        final addrType = data[3];

        String address;
        int port = 0;

        switch (addrType) {
          case 0x01: // IPv4
            address = data.sublist(4, 8).join('.');
            port = (data[8] << 8) | data[9];
            break;
          case 0x04: // IPv6
            address = '::1';
            port = (data[16] << 8) | data[17];
            break;
          case 0x03: // Domain name
            final domainLength = data[4];
            address = String.fromCharCodes(data.sublist(5, 5 + domainLength));
            port = (data[5 + domainLength] << 8) | data[6 + domainLength];
            break;
          default:
            log.warning('Unknown address type: $addrType');
            client.close();
            return;
        }

        try {
          remoteSocket = await Socket.connect(address, port,
              timeout: Duration(seconds: 5));
          print('Connected to $address:$port');

          // Reply: succeeded
          client.add(Uint8List.fromList([5, 0, 0, 1, 0, 0, 0, 0, 0, 0]));
          await client.flush();
        } catch (e) {
          print('Connection to $address:$port failed: $e');
          client.close();
        }
        currentState = States.proxying;
        startProxy();
        break;

      case States.proxying:
        try {
          remoteSocket?.add(data);
        } catch (e) {
          print('[-] failed: $e');
          client.close();
        }

        break;
    }
  }

  String parseAddress(List<int> data, int addrType) {
    switch (addrType) {
      case 0x01: // IPv4
        return data.sublist(4, 8).join('.');
      case 0x04: // IPv6
        return '::1';
      case 0x03: // Domain name
        final domainLength = data[4];
        return String.fromCharCodes(data.sublist(5, 5 + domainLength));
      default:
        throw Exception('Unknown address type: $addrType');
    }
  }
}

// Handler for TCP transfer and relaying data
class TcpTransferHandler {
  final String address;
  final Addr addr;
  final int port;
  late final Socket remSocket;

  TcpTransferHandler(this.address, this.addr, this.port);

  Future<void> connectAndRelay({required Socket clientSocket}) async {
    try {
      Socket remSocket = await _connectToRemote();
      _sendSocks5Reply(clientSocket, remSocket);

      _relayTraffic(clientSocket, remSocket);
    } catch (e) {
      log.severe('Connection failed: $e');
      clientSocket.close();
    }
  }

  Future<Socket> _connectToRemote() async {
    late Socket remSocket;
    switch (addr) {
      case Addr.v4:
        remSocket = await Socket.connect(address, port);
        break;
      case Addr.v6:
        remSocket = await Socket.connect(address, port);
        break;
      case Addr.domain:
        remSocket = await Socket.connect(address, port);
        break;
      default:
        throw Exception('Unsupported address type');
    }
    return remSocket;
  }

  void _sendSocks5Reply(Socket clientSocket, Socket remSocket) {
    final reply = BytesBuilder()
      ..addByte(5) // SOCKS version
      ..addByte(0) // Connection succeeded
      ..addByte(0) // Reserved

      ..addByte(remSocket.address.type == InternetAddressType.IPv4
          ? 1
          : 4) // Address type

      ..add(remSocket.address.rawAddress) // Address
      ..addByte(remSocket.port >> 8) // Port (high byte)
      ..addByte(remSocket.port & 0xFF); // Port (low byte)

    clientSocket.add(reply.toBytes());
  }

  void _relayTraffic(Socket clientSocket, Socket remSocket) {
    clientSocket.listen(
          (data) {
        remSocket.add(data);
      },
      onError: (error) {
        log.severe('Client socket error: $error');
        remSocket.close();
      },
      onDone: () {
        remSocket.close();
      },
    );
    print('[+]TcpTransferHandler _relayTraffic');
    // Listen for data from the remote socket and forward it to the client socket
    remSocket.listen(
          (data) {
        clientSocket.add(data);
      },
      onError: (error) {
        log.severe('Remote socket error: $error');
        clientSocket.close();
      },
      onDone: () {
        clientSocket.close();
      },
    );
  }
}

enum States { handshake, handling, proxying }
